chainer/optimizer_hooks/gradient_clipping.py

```diff
@@ -5,22 +5,52 @@
 
 import chainer
 from chainer import backend
-from chainer import cuda
-
-
-def _sum_sqnorm(arr):
-    sq_sum = collections.defaultdict(float)
-    for x in arr:
-        with cuda.get_device_from_array(x) as dev:
-            x = x.ravel()
-            s = x.dot(x)
-            sq_sum[int(dev)] += s
-    # If only a single device is used, aggregate square norms on it.
-    if len(sq_sum) == 1:
-        with cuda.get_device_from_array(arr[0]):
-            return sum(six.itervalues(sq_sum))
+
+
+def _sum_sqnorm_grads(params):
+    # Calculates sum of squares of gradients.
+
+    # Returns a tuple of the sum and the device of the sum.
+    # The device will be `None` in multi-device case.
+
+    # If the inputs are on a single device, the sum is returned
+    # as an ndarray on the device, so that no synchronization is taken place.
+    # If there are multiple devices, accumulation is done on each device
+    # first, and the total sum is returned as a python float.
+
+    assert len(params) > 0
+
+    params_grouped = collections.defaultdict(list)
+    devices_map = {}
+
+    # Group params by devices.
+    for param in params:
+        device = param.device
+        params_grouped[device.name].append(param)
+        devices_map[device.name] = device
+
+    # Calculates partial sums for each device.
+    sq_sums = []
+    for device_name, paramlist in six.iteritems(params_grouped):
+        device = devices_map[device_name]
+        with chainer.using_device(device):
+            dots = []
+            for param in paramlist:
+                g = param.grad
+                g = g.ravel()
+                dots.append(g.dot(g))
+            sq_sums.append(sum(dots))
```

Comment:
Just curious but is it faster to utilize an append to list -> reduce style rather than having a running sum (to me a running sum would be the first go-to) for `dots`?

Reply:
I don't think there's a perceivable difference on performance.
I think it's simpler. `sum` could be replaced with `xp.sum` if it supports a list argument (https://github.com/cupy/cupy/issues/2280)